crates/sui-core/src/authority_aggregator.rs

```diff
@@ -1458,20 +1518,83 @@ where
+        authorities: &BTreeSet<AuthorityName>,
+        timeout_total: Option<Duration>,
+    ) -> SuiResult<CheckpointResponse> {
+        self.quorum_once_with_timeout(
+            &[],
+            Some(authorities),
+            |_, client| Box::pin(async move { client.handle_checkpoint(request.clone()).await }),
+            self.timeouts.serial_authority_request_timeout,
+            timeout_total,
+        )
+        .await
+    }
+
+    pub async fn get_certified_checkpoint(
+        &self,
+        request: &CheckpointRequest,
+        // authorities known to have the checkpoint we are requesting.
+        authorities: &BTreeSet<AuthorityName>,
+        timeout_total: Option<Duration>,
+    ) -> SuiResult<(CertifiedCheckpointSummary, Option<CheckpointContents>)> {
+        self.quorum_once_with_timeout(
+            &[],
+            Some(authorities),
+            |_, client| {
+                Box::pin(async move {
+                    let resp = client.handle_checkpoint(request.clone()).await?;
+
+                    if let CheckpointResponse {
+                        info:
+                            AuthorityCheckpointInfo::Past(AuthenticatedCheckpoint::Certified(past)),
+                        detail,
+                    } = resp
+                    {
+                        Ok((past, detail))
+                    } else {
+                        Err(SuiError::GenericAuthorityError {
+                            error: "expected Certified checkpoint".into(),
+                        })
+                    }
+                })
+            },
+            self.timeouts.serial_authority_request_timeout,
+            timeout_total,
+        )
+        .await
+    }
+
     pub async fn handle_transaction_and_effects_info_request(
         &self,
         digests: &ExecutionDigests,
-    ) -> Result<TransactionInfoResponse, SuiError> {
+        // authorities known to have the effects we are requesting.
+        authorities: &BTreeSet<AuthorityName>,
+        timeout_total: Option<Duration>,
+    ) -> SuiResult<(CertifiedTransaction, SignedTransactionEffects)> {
         self.quorum_once_with_timeout(
-            |_name, client| {
+            &[],
+            Some(authorities),
+            |authority, client| {
                 Box::pin(async move {
-                    client
+                    let resp = client
                         .handle_transaction_and_effects_info_request(digests)
-                        .await
+                        .await?;
+
+                    match (resp.certified_transaction, resp.signed_effects) {
+                        (Some(cert), Some(effects)) => Ok((cert, effects)),
+                        // The caller is passing in authorities that have claimed to have the cert and
+                        // effects, so if they now say they don't, they're byzantine.
+                        _ => Err(SuiError::ByzantineAuthoritySuspicion { authority }),
```

Comment:
I wonder if this is necessary. I would also love to be able to use this function for a general case where I don't know who would have a specific cert/effect.

Reply:
I am currently working on fetching/executing certs for checkpoints - in that case we won't know precisely who has the certs, so I will have to do something along the lines you suggest soon anyway.